### DISCUSSION TITLE
Allow specifying the exact compiler/toolset version in a triples

### DISCUSSION THREAD

**Discussion by gix** [microsoft/vcpkg#17815] [Ideas]
Triplets may contain `VCPKG_PLATFORM_TOOLSET` to set the compiler toolchain to use, but this option is too coarse as it only allows v140, v141 and v142. In MSBuild or with VSDevCmd I can choose the exact version of the toolchain, e.g. 14.24.28314.

There should be a way to set the exact version, either using `VCPKG_PLATFORM_TOOLSET` or another variable.

Why is this important? Because AFAICT from the docs static linking of the runtime requires and whole program optimization requires the exact version for all libraries. An application built with a specific toolset may not link against vcpkg packates built with a newer toolchain.

And this is not theoretical, since I just recently had the case where a library built with vcpkg (toolset `14.25.28610` chosen automatically) failed to link with an application using `14.24.28314`, as the required `__dyn_tls_on_demand_init` symbol is not in the older `libcmt.lib`.

**Comment by LilyWangL** [comment]
Thanks for posting this issue. Closed as duplicate of #9139.

**Comment by gix** [comment]
#9139 doesn't look like a duplicate, it merely allows setting `VCPKG_PLATFORM_TOOLSET` via environment variable. But as explained above, `VCPKG_PLATFORM_TOOLSET` is not sufficient enough in its current state.

**Comment by MVoz** [comment]
https://github.com/microsoft/vcpkg/blob/644ba5d998fec302fc9936a27f3be95f085ba62d/toolsrc/src/vcpkg/visualstudio.cpp#L254

it is strictly registered, you can't change it, reassign it only

attempt

`cmake -T""`

`set(VS_PLATFORM_TOOLSET "host=x64,version=14.24.28314")`

https://github.com/microsoft/vcpkg/pull/11140

**Comment by MVoz** [comment]
try adding an option

example msbuild 

```
    <VCToolsVersion>14.25.28610</VCToolsVersion>
    <WindowsTargetPlatformVersion>10.0.19582.0</WindowsTargetPlatformVersion>
    <PlatformToolset>v142</PlatformToolset>
```

`msbuild /p:PlatformToolset=v142 /p:WindowsTargetPlatformVersion=10.0.19582.0 /p:VCToolsVersion=14.25.28610`


add VCToolsVersion

https://github.com/microsoft/vcpkg/blob/6fd7e9c7120c2faf6846e315ac8073dd04367059/scripts/cmake/vcpkg_install_msbuild.cmake#L132

**Comment by Chronial** [comment]
+1 To this feature request.

We pin an exact compiler version in our builds to make sure that all devs in our team actually built the same thing. As far as I understand, this is currently not possible with vcpkg. Our own code is always built with the exact toolchain version we specify. But the vcpkg packages are built with whichever version happens to be the most recent on the developers machine.

This is rather problematic for us and will probably force us to never run vcpkg on dev machines, but requires us to have a separate build machine that only has the correct msvc toolset version installed.

**Comment by Chronial** [comment]
I created a PR for this: https://github.com/microsoft/vcpkg-tool/pull/199

**Reply by Chronial** [reply]
Merged 🥳
